src/dev/lupio/lupio_sys.cc

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2021 The Regents of the University of California
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_sys.hh"
+
+#include "debug/LupioSYS.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioSYS.hh"
+#include "sim/sim_exit.hh"
+
+namespace gem5
+{
+
+LupioSYS::LupioSYS(const Params &params) :
+    BasicPioDevice(params, params.pio_size)
+{
+    DPRINTF(LupioSYS, "LupioSYS initalized\n");
+}
+
+uint8_t
+LupioSYS::lupioSYSRead(uint8_t addr)
+{
+    switch (addr >> 2) {
+        case LUPIO_SYS_HALT:
+            DPRINTF(LupioSYS, "Trying to halt\n");
+            exitSimLoopNow("LUPIO_SYS_HALT called, exiting", 0, 0, false);
+            break;
+        case LUPIO_SYS_REBT:
+            DPRINTF(LupioSYS, "Trying to reboot\n");
+            exitSimLoopNow("LUPIO_SYS_REBT called, exiting", 0, 0, false);
+            break;
+
+        default:
+            panic("Unexpected write to the LupioRTC device at address %d!",
+                    addr);
+            break;
+    }
+    return 0;
```

Comment:
So, for the read, and according to the fixed specs, you just want to return `0`. The actual exit/reboot should only be if the registers are being written to.